RPC: allow binding of restricted port in addition to core port

Timo614 opened the pull request:
This branch adds two new configuration options `--rpc-restricted-bind-port` and `--testnet-rpc-restricted-bind-port` which allow for a second port to be opened to handle RPC requests.

Let me know if this direction is unwise. I spent some time looking over the code and this was the approach that seemed most sensible to me but I'm still learning C++ so completely understand if I've done something asinine. Currently the login is the same across both but if there's value that could be split into a separate set of configuration arguments as well.

I was able to test this by using the `/getinfo` call across both and confirmed that `/get_peer_list` only works for the unrestricted. 

It's currently possible for someone to both set the core RPC port as restricted via `--restricted-rpc` and enable the restricted RPC port -- if it's worth it I can add something quickly to prevent that though.

ref: #835

Timo614 commented on src/daemon/rpc.h:
```diff
@@ -47,35 +47,41 @@ class t_rpc final
+    if (!m_server.init(vm, restricted, testnet, port))
     {
-      throw std::runtime_error("Failed to initialize core rpc server.");
+      throw std::runtime_error("Failed to initialize " + m_description + " rpc server.");
```
I wasn't able to use the `<<` syntax here as it returned an error during compilation related to `invalid operands to binary expression ('const char *' and 'const std::string'...)`

Trying to figure out why (and will replace this once I do figure that out) but just making note of it if anyone knows offhand (still learning C++ so this isn't immediately obvious to me).

moneromooo-monero replied:
<< is when the object is something which has this operator, such as a stringtream. operator+ is the right thing to use here.

moneromooo-monero commented on src/daemon/daemon.cpp:
```diff
@@ -60,19 +60,27 @@ struct t_internals {
+    const auto restricted_rpc_port = command_line::get_arg(vm, testnet ? cryptonote::core_rpc_server::arg_testnet_rpc_restricted_bind_port : cryptonote::core_rpc_server::arg_rpc_restricted_bind_port);
+    rpcs.emplace_back(new t_rpc{vm, core, p2p, restricted, testnet, main_rpc_port, "core"});
+
+    if(restricted_rpc_port.size())
```
That should probably be somhting like "if (command_line::is_arg_defaulted(vm, ...))" and have the get_arg inside the if.

moneromooo-monero commented on src/daemon/daemon.cpp:
```diff
@@ -135,14 +143,15 @@ bool t_daemon::run(bool interactive)
 
-    std::unique_ptr<daemonize::t_command_server> rpc_commands;
+    for(const auto& rpc: mp_internals->rpcs)
+      rpc->run();
```
Logically, it seems odd to call run on a const object. I'd make it non const, it's less surprising I think ?

Timo614 replied:
Sounds good, yeah agreed I'll switch those instances of it to remove that.

moneromooo-monero commented on src/daemon/rpc.h:
```diff
@@ -47,35 +47,41 @@ class t_rpc final
+    , const bool restricted
+    , const bool testnet
+    , const std::string port
+    , const std::string description
```
Maybe forgot a & ? (unless you know what you're doing with copy elision, I'm not very familiar with those)

Timo614 commented:
I've pushed up some changes via https://github.com/Timo614/monero/commit/a45022eda85a048707910e14185fb1b5328d2a11 which I squashed and force pushed up here to one commit.

- Added the `&` to the method signatures for port and description to pass by reference
- Switched the iteration of the `rpcs` to remove the `const` declaration
- Used `is_arg_defaulted` for confirming whether the restricted port was requested

Timo614 commented:
Going to push up one more change to just rename a variable I introduced in https://github.com/Timo614/monero/commit/a45022eda85a048707910e14185fb1b5328d2a11 from `restricted_rpc_arg` to `restricted_rpc_port_arg` as there's another `restricted-rpc` arg and I want to remove confusion.

hyc commented:
IMO we've got a useless proliferation of duplicate options, for testnet vs mainnet. It's not as if we allow a single daemon to operate on both simultaneously (do we?)

There should be just one `--testnet` option, which completely changes the daemon's operating context, and thus all of the other parameters, in one fell swoop.

Timo614 commented:
That's true.

For the purposes of this branch I could remove the testnet specific one in favor of just checking the other argument regardless of testnet specification.

For `--testnet-rpc-bind-port` should I leave that logic in place? I'd worry about breaking any scripts that may already be passing that in.

I guess one option could be to remove the default value for the `--testnet-rpc-bind-port` keep the option but then have the logic just fallback to the `--rpc-bind-port` if it's not passed in. It currently defaults to the value of the `RPC_DEFAULT_PORT` constant (same as `--rpc-bind-port`) so that would at least keep the behavior consistent across the normal and restricted ports without breaking the existing use case.

So it'll be kind of a deprecated option in that case that we still support but don't encourage usage of? May be worth in the future adding some flag on it to hide it from the output of help.

Or does it not matter compatibility wise since it's just a testnet argument so if it's removed it really just breaks a very specific use case?

Timo614 commented:
I pushed up a change: https://github.com/timo614/monero/commit/6201ccd55871d0558990c2c5a11566660774ba08 which changes the logic to reflect the review feedback. 

I removed the `--testnet-rpc-restricted-bind-port` in favor of always using the `--rpc-restricted-bind-port`. 

I also modified the existing logic for `--testnet-rpc-bind-port` to no longer default to the `RPC_DEFAULT_PORT` constant. Instead callers now check to see if that's present and if so use it otherwise will fall back on `--rpc-bind-port`. Users who don't pass in a bind port still use the same port as they had previously as the `--rpc-bind-port` argument defaults to that. 

I mentioned it in my above comment but the one case that may have issues is just if someone previously was passing in `--testnet` and a value for `--rpc-bind-port` but no value for `--testnet-rpc-bind-port` as in that case it would have defaulted to `RPC_DEFAULT_PORT` previously but now will use the value from `--rpc-bind-port`.

I'll rebase and squash this.

Timo614 commented:
Upon further review (outside of this logic) there are other testnet arguments currently being used such as `--testnet-data-dir` and `--zmq-testnet-rpc-bind-port` I could make modifications for them in this branch but it seems out of scope. 

I'll take a look at what would need to be changed there but it may be worth having those modified in a separate pull request as they aren't really altered here by these changes.

moneromooo-monero commented:
I'd rather the testnet arguments merge be in a totally separate patch (PR or not at your convenience, I don't really care, just a separate patch as it's logically separate).

moneromooo-monero commented:
BTW, I think fluffypony had a preference for separating the arguments, so I'll go ping him to see this.

Timo614 commented:
I pushed up https://github.com/Timo614/monero/commit/7c1a19a77f9fb73db75466e4cd758b158fcf7437

Which reverts https://github.com/Timo614/monero/commit/6201ccd55871d0558990c2c5a11566660774ba08

I'll hold off on squashing it for now. If we decide to combine the testnet and regular arguments I'll nix the testnet one here and save any additional changes to combine the testnet and regular versions to outside of this branch (the behavior of the `--rpc-bind-port` and `--rpc-restricted-bind-port` will diverge in that case for testnet but only until the additional PR is merged so not terrible).

Timo614 commented:
I ended up squashing this -- it's back to the code it was at previously where there's the two RPC arguments (one for testnet and one for the normal use case). Can swing back and remove the testnet one for this particular use case if we decide to combine them or can just do that in a subsequent patch and deprecate that command in a similar way (or just delete it if we haven't released a new version yet / wouldn't break backwards compatibility).
